helix-core/src/main/java/org/apache/helix/controller/changedetector/ResourceChangeDetector.java

```diff
@@ -0,0 +1,164 @@
+  private Map<String, ? extends HelixProperty> _newPropertyMap;
+
+  // The following caches the computation results
+  private Map<HelixConstants.ChangeType, Collection<String>> _changedItems = new HashMap<>();
+  private Map<HelixConstants.ChangeType, Collection<String>> _addedItems = new HashMap<>();
+  private Map<HelixConstants.ChangeType, Collection<String>> _removedItems = new HashMap<>();
+
+  /**
+   * Compare the underlying HelixProperty objects and produce a collection of names of changed
+   * properties.
+   * @return
+   */
+  private Collection<String> getChangedItems() {
+    Collection<String> changedItems = new HashSet<>();
+    _oldPropertyMap.forEach((name, property) -> {
+      if (_newPropertyMap.containsKey(name) && !property.equals(_newPropertyMap.get(name))) {
+        changedItems.add(name);
+      }
+    });
+    return changedItems;
+  }
+
+  /**
+   * Return a collection of names that are newly added.
+   * @return
+   */
+  private Collection<String> getAddedItems() {
+    Collection<String> addedItems = new HashSet<>(_newPropertyMap.keySet());
+    addedItems.removeAll(_oldPropertyMap.keySet());
+    return addedItems;
+  }
+
+  /**
+   * Return a collection of names that were removed.
+   * @return
+   */
+  private Collection<String> getRemovedItems() {
+    Collection<String> removedItems = new HashSet<>(_oldPropertyMap.keySet());
+    removedItems.removeAll(_newPropertyMap.keySet());
+    return removedItems;
+  }
+
+  /**
+   * Based on the change type given, call the right getters for two property maps and save the
+   * references as oldPropertyMap and newPropertyMap.
+   * @param changeType
+   */
+  private void determinePropertyMapByType(HelixConstants.ChangeType changeType) {
+    switch (changeType) {
+    case INSTANCE_CONFIG:
+      _oldPropertyMap = _oldCache.getInstanceConfigMap();
+      _newPropertyMap = _newCache.getInstanceConfigMap();
+      break;
+    case IDEAL_STATE:
+      _oldPropertyMap = _oldCache.getIdealStateMap();
+      _newPropertyMap = _newCache.getIdealStateMap();
+      break;
+    case RESOURCE_CONFIG:
+      _oldPropertyMap = _oldCache.getResourceConfigMap();
+      _newPropertyMap = _newCache.getResourceConfigMap();
+      break;
+    case LIVE_INSTANCE:
+      _oldPropertyMap = _oldCache.getLiveInstances();
+      _newPropertyMap = _newCache.getLiveInstances();
+      break;
+    default:
+      throw new HelixException(String
+          .format("ResourceChangeDetector does not support the given change type: %s", changeType));
+    }
+  }
+
+  /**
+   * Makes the current newCache the oldCache and reads in the up-to-date cache for change
+   * computation. To be called in the controller pipeline.
+   * @param dataProvider newly refreshed DataProvider (cache)
+   */
+  public synchronized void updateCache(ResourceControllerDataProvider dataProvider) {
+    // Do nothing if nothing has changed
+    if (dataProvider.getRefreshedChangeTypes().isEmpty()) {
```

Comment:
If nothing has been changed, shall we still clean up all the caches here?

Reply:
Yes. Good catch.